Add LedgerBalances Indicator

esoergel opened the pull request:
This is working end-to-end now, so feel free to test it out, but I'm looking for feedback at this time.

esoergel committed:
Make basic, failing test for ledger_balances

esoergel committed:
Build basic LedgerBalances indicator

That actually gets pretty far through the tests

esoergel committed:
Mock out StockState.objects.get

esoergel committed:
Actually look up StockState from db

esoergel committed:
Actually use product_code

esoergel commented:
@benrudolph @czue

esoergel commented on corehq/apps/userreports/indicators/factory.py:
```diff
@@ -85,6 +87,42 @@ def _construct_column(choice):
+                id=_construct_column(product_code),
+                datatype="decimal",
+            ),
+            getter=_make_get_stock_state(product_code),
```
I saw some other indicator types used a`getter` property on the `IndicatorSpec`.  The problem is this needs to know the product_code, and I'm not sure how else to pass it through.  The other thing I'm thinking of (but haven't yet implemented) is reducing the db hits here - a simple optimization is:
```python
# switch from
StockState.objects.get(
    section_id=wrapped_spec.ledger_section,
    case_id=doc['_id'],
    sql_product__code=product_code,
)
# to
StockState.objects.filter(
    section_id=wrapped_spec.ledger_section,
    case_id=doc['_id'],
    sql_product__code__in=product_codes,
)
```
Such an optimization would require passing in shared state to the getter like I've done here with `product_code`.

czue replied:
agree that optimization would begood

czue commented:
build is failing
```
======================================================================
ERROR: test_ledger_balances_indicator (corehq.apps.userreports.tests.test_indicators.LedgerBalancesIndicatorTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/mock.py", line 1224, in patched
    return func(*args, **keywargs)
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/userreports/tests/test_indicators.py", line 489, in test_ledger_balances_indicator
    values = indicator.get_values(doc, EvaluationContext(doc, 0))
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/userreports/indicators/__init__.py", line 97, in get_values
    return [val for ind in self.indicators for val in ind.get_values(item, context)]
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/userreports/indicators/__init__.py", line 82, in get_values
    return [ColumnValue(self.column, self.getter(item, context))]
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/userreports/indicators/factory.py", line 107, in get_stock_state
    sql_product__code=product_code,
TypeError: mock_getter() got an unexpected keyword argument 'sql_product__code'
----------------------------------------------------------------------
```

czue commented on corehq/apps/userreports/indicators/factory.py:
```diff
@@ -85,6 +87,42 @@ def _construct_column(choice):
+            try:
+                return StockState.objects.get(
+                    section_id=wrapped_spec.ledger_section,
+                    case_id=doc['_id'],
```
I think it would be nice if this was an expression, so that you could link a ledger row to something other than the primary case table containing that ledger.

esoergel replied:
:+1:

czue commented on corehq/apps/userreports/indicators/factory.py:
```diff
@@ -85,6 +87,42 @@ def _construct_column(choice):
+    def _construct_column(product):
+        return '{col}_{product}'.format(col=spec['column_id'], product=product)
+
+    def _make_get_stock_state(product_code):
```
minor, but I would change this to `get_current_stock_on_hand` or similar for clarity

benrudolph replied:
+1

czue commented on corehq/apps/userreports/indicators/factory.py:
```diff
@@ -85,6 +87,42 @@ def _construct_column(choice):
+            getter=_make_get_stock_state(product_code),
+        ) for product_code in wrapped_spec.product_codes
+    ]
+    return CompoundIndicator(base_display_name, product_indicators)
```
Given that this logic is fairly complex and specific to the product/ledger models, I'm curious whether you considered subclassing `ConfigurableIndicator` directly, rather than using `CompoundIndicator`, and if so why you landed on this approach.

esoergel replied:
I didn't consider that, actually - I was mostly copying the example of the choice list indicator.  Looking at it now, it looks like that would make sense - particularly for shared queries, as you mentioned, I've been looking for a better place for that to live.

czue commented:
looking good! some questions but seems close!

benrudolph commented on corehq/apps/userreports/tests/test_indicators.py:
```diff
@@ -438,3 +441,54 @@ def testDecimal(self):
+        else:
+            raise StockState.DoesNotExist()
+
+    @patch('corehq.apps.userreports.indicators.factory.StockState.objects')
```
nice

esoergel replied:
The patch is a bit too closely tied to the implementation for my liking (see also: current test failure), but I figure it's okay as long as the query isn't much more complicated than passing a couple parameters to `StockState.objects.get`

esoergel committed:
Switch tests to use product code

esoergel committed:
Move ledger stuff to new LedgerBalancesIndicator

(and tests pass!)

dimagimon commented on corehq/apps/userreports/indicators/factory.py:
```diff
@@ -3,14 +3,22 @@
+    LedgerBalancesIndicatorSpec,
+    RawIndicatorSpec,
 )
+from corehq.apps.commtrack.models import StockState
```
F401 'StockState' imported but unused

czue commented on corehq/apps/userreports/indicators/__init__.py:
```diff
@@ -95,3 +96,34 @@ def get_columns(self):
+    def get_values(self, item, context=None):
+        case_id = item['_id']
+        return [ColumnValue(self._make_column(product_code),
+                            self._get_value(case_id, product_code))
```
I think you can run a single query first and then just use the results of that query rather than hitting the DB once for each product, no?

esoergel replied:
Yup, that's what I'm planning on doing, I was just moving stuff over to the indicator and hadn't pulled that out yet.

esoergel committed:
Switch to bulk StockStates query

esoergel committed:
Test db component that I'd mocked out

It seems preposterous that it takes this much setup to test that one little
query.

esoergel committed:
Add case_id_expression to ledger balance indicator

esoergel commented on corehq/apps/userreports/indicators/specs.py:
```diff
@@ -97,3 +97,18 @@ class ChoiceListIndicatorSpec(PropertyReferenceIndicatorSpecBase):
+            "type": "property_name",
+            "property_name": "_id",
+        }
+        return ExpressionFactory.from_spec(expression_spec)
```
I figured since this will most likely always be this same expression, it makes sense to not require you to specify.  Happy to hear arguments if you disagree.

czue replied:
I think I disagree. My general philosophy for UCR has to push all the config into the json and not have any surprising magic happen in the framework. If we want to provide a default I'd advocate for that default being specified either in the UI or on save, so that everything that deals with data is fully specified in the self-contained definition.

esoergel replied:
I can move it to the `wrap` method if you think that makes sense.  Could also just drop if that doesn't fit with the style of this module.

esoergel replied:
Just tried out moving it to the `wrap` method - it looks like when editing a data source config, the indicators are wrapped during validation, but the wrapped results aren't saved, so moving this to `wrap` would require some reworking of that initial save.

Since having defaults is inconsistent with most of the UCR stuff, I just dropped it.

esoergel commented on corehq/apps/userreports/tests/test_indicators.py:
```diff
@@ -438,3 +447,83 @@ def testDecimal(self):
+        )
+
+
+class TestGetValuesByProduct(TestCase):
```
I mocked out the `_get_values_by_product` query in the indicator test, so this tests that explicitly.  It's mostly tedious setup work.

czue replied:
> It's mostly tedious setup work.

sounds like commtrack :disappointed:

esoergel commented:
This should be g2g now.

esoergel committed:
Add ledger balances indicator to UCR docs

[ci skip]

czue commented on corehq/apps/userreports/indicators/__init__.py:
```diff
@@ -95,3 +97,34 @@ def get_columns(self):
+        values_by_product = StockState.objects.filter(
+            section_id=ledger_section,
+            case_id=case_id,
+        ).values_list('sql_product__code', 'stock_on_hand')
```
any reason to not also filter by product code here?

esoergel replied:
nope, will update

czue commented on corehq/apps/userreports/indicators/specs.py:
```diff
@@ -97,3 +97,18 @@ class ChoiceListIndicatorSpec(PropertyReferenceIndicatorSpecBase):
+class LedgerBalancesIndicatorSpec(IndicatorSpecBase):
+    type = TypeProperty('ledger_balances')
+    product_codes = ListProperty(required=True)
+    ledger_section = StringProperty()
```
should this also be required?

esoergel replied:
:+1:

esoergel committed:
Filter by product_codes and require ledger section

esoergel committed:
Require case_id_expression and get rid of default

czue commented:
looks good - thanks for all the back and forth!

esoergel commented:
awesome! thanks for all the feedback

benrudolph commented:
:tada:
